File: go/pkg/chatgear/client_port.go
package chatgear

import (
	"context"
	"fmt"
	"sync"
	"time"

	"github.com/haivivi/giztoy/pkg/audio/opusrt"
)

// ClientPort implements ClientPortTx and ClientPortRx as a bidirectional audio port.
// It manages audio input encoding/sending and output buffering/command receiving.
type ClientPort struct {
	tx      UplinkTx
	context context.Context
	cancel  context.CancelFunc
	logger  Logger

	// Output - audio from server (ClientPortRx)
	playBuffer *opusrt.RealtimeBuffer
	commands   chan *SessionCommandEvent
	commandsMu sync.Mutex // protects commands channel and closed flag
	closed     bool

	// Input - audio to server (ClientPortTx)
	recBuffer    *opusrt.RealtimeBuffer
	inputStamp   time.Time
	inputStampMu sync.Mutex
}

// NewClientPort creates a new ClientPort for the given UplinkTx.
func NewClientPort(ctx context.Context, tx UplinkTx) *ClientPort {
	ctx, cancel := context.WithCancel(ctx)
	p := &ClientPort{
		tx:      tx,
		context: ctx,
		cancel:  cancel,
		logger:  DefaultLogger(),

		playBuffer: opusrt.RealtimeFrom(opusrt.NewBuffer(2 * time.Minute)),
		commands:   make(chan *SessionCommandEvent, 32),
		recBuffer:  opusrt.RealtimeFrom(opusrt.NewBuffer(2 * time.Minute)),
	}
	go p.streamingInputLoop()
	return p
}

// --- ClientPortTx (sending to server) ---

// SendOpusFrames sends a stamped opus frame to the server.
func (p *ClientPort) SendOpusFrames(stampedOpusFrame []byte) error {
	frame, stamp, ok := opusrt.FromStamped(stampedOpusFrame)
	if !ok {
		return fmt.Errorf("chatgear: invalid stamped opus frame")
	}
	return p.tx.SendOpusFrames(p.context, stamp, frame)
}

// SendState sends a state event to the server.
func (p *ClientPort) SendState(state *GearStateEvent) error {
	return p.tx.SendState(p.context, state)
}

// SendStats sends a stats event to the server.
func (p *ClientPort) SendStats(stats *GearStatsEvent) error {
	return p.tx.SendStats(p.context, stats)
}

// --- ClientPortRx (receiving from server) ---

// Frame reads the next Opus frame from the server.
// Implements opusrt.FrameReader.
func (p *ClientPort) Frame() (opusrt.Frame, time.Duration, error) {
	return p.playBuffer.Frame()
}

// Commands returns a channel that receives commands from the server.
func (p *ClientPort) Commands() <-chan *SessionCommandEvent {
	return p.commands
}

// --- Handle (input from transport) ---

// HandleOpusFrames handles incoming Opus frames from the server.
func (p *ClientPort) HandleOpusFrames(stampedOpusFrame []byte) {
	if _, err := p.playBuffer.Write(stampedOpusFrame); err != nil {
		p.logger.ErrorPrintf("handle opus frames: %v", err)
	}
}

// HandleCommand handles an incoming command from the server.
func (p *ClientPort) HandleCommand(cmd *SessionCommandEvent) {
	p.commandsMu.Lock()
	defer p.commandsMu.Unlock()

	if p.closed {
		return
	}

	select {
	case p.commands <- cmd:
	default:
		p.logger.WarnPrintf("commands channel is full, drop command")
	}
}

// WriteRecordingFrame writes a stamped opus frame to be sent to the server.
func (p *ClientPort) WriteRecordingFrame(stampedOpusFrame []byte) {
	if _, err := p.recBuffer.Write(stampedOpusFrame); err != nil {
		p.logger.ErrorPrintf("write recording frame: %v", err)
	}
}

// --- Lifecycle ---

// Context returns the port's context.
func (p *ClientPort) Context() context.Context {
	return p.context
}

// Close closes the port.
func (p *ClientPort) Close() error {
	p.cancel()
	p.playBuffer.Close()
	p.recBuffer.Close()

	// Safely close channel under lock to prevent send-to-closed-channel panic
	p.commandsMu.Lock()
	p.closed = true
	close(p.commands)
	p.commandsMu.Unlock()

	return p.tx.Close()
}

// RecvFrom receives data from the given DownlinkRx until closed or error.
// This method blocks; use `go port.RecvFrom(rx)` for non-blocking operation.
// Returns the first error encountered, or nil if all iterators completed normally.
func (p *ClientPort) RecvFrom(rx DownlinkRx) error {
	defer rx.Close()

	var wg sync.WaitGroup
	var mu sync.Mutex
	var firstErr error

	setErr := func(err error) {
		if err == nil {
			return
		}
		mu.Lock()
		if firstErr == nil {
			firstErr = err
		}
		mu.Unlock()
	}

	wg.Add(2)
	go func() {
		defer wg.Done()
		for frame, err := range rx.OpusFrames() {
			if err != nil {
				setErr(err)
				return
			}
			p.HandleOpusFrames(frame)
		}
	}()
	go func() {
		defer wg.Done()
		for cmd, err := range rx.Commands() {
			if err != nil {
				setErr(err)
				return
			}
			p.HandleCommand(cmd)
		}
	}()
	wg.Wait()
	return firstErr
}

func (p *ClientPort) streamingInputLoop() {
	for {
		select {
		case <-p.context.Done():
			return
		default:
		}

		frame, _, err := p.recBuffer.Frame()
		if err != nil {
			p.logger.ErrorPrintf("read frame from rec buffer: %v", err)
			return
		}

		if len(frame) == 0 {
			continue
		}

		p.inputStampMu.Lock()
		now := time.Now()
		if p.inputStamp.Before(now) {
			p.inputStamp = now
		}
		stamp := p.inputStamp
		p.inputStamp = p.inputStamp.Add(frame.Duration())
		p.inputStampMu.Unlock()

		if err := p.tx.SendOpusFrames(p.context, opusrt.FromTime(stamp), frame); err != nil {
			p.logger.ErrorPrintf("send opus frame: %v", err)
			continue
		}
	}
}

File: rust/chatgear/src/client_port.rs
//! Client port implementation.
//!
//! ClientPort manages client-side audio input/output and command handling.

use crate::{
    ClientPortRx, ClientPortTx, DownlinkRx, GearStateEvent, GearStatsEvent,
    PortError, SessionCommandEvent, StampedOpusFrame, UplinkTx,
};
use async_trait::async_trait;
use std::sync::Arc;
use tokio::sync::{mpsc, Mutex, RwLock};
use tokio_util::sync::CancellationToken;
use tracing::warn;

/// Client port for managing audio input/output and commands.
pub struct ClientPort<T: UplinkTx> {
    tx: Arc<T>,
    cancel: CancellationToken,

    // Output - audio from server (ClientPortRx)
    opus_frames_rx: Arc<Mutex<mpsc::Receiver<StampedOpusFrame>>>,
    opus_frames_tx: mpsc::Sender<StampedOpusFrame>,
    commands_rx: Arc<Mutex<mpsc::Receiver<SessionCommandEvent>>>,
    commands_tx: mpsc::Sender<SessionCommandEvent>,

    closed: Arc<RwLock<bool>>,
}

impl<T: UplinkTx + 'static> ClientPort<T> {
    /// Creates a new ClientPort.
    pub fn new(tx: T) -> Self {
        let (opus_tx, opus_rx) = mpsc::channel(1024);
        let (cmd_tx, cmd_rx) = mpsc::channel(32);

        Self {
            tx: Arc::new(tx),
            cancel: CancellationToken::new(),
            opus_frames_rx: Arc::new(Mutex::new(opus_rx)),
            opus_frames_tx: opus_tx,
            commands_rx: Arc::new(Mutex::new(cmd_rx)),
            commands_tx: cmd_tx,
            closed: Arc::new(RwLock::new(false)),
        }
    }

    /// Returns the cancellation token.
    pub fn cancellation_token(&self) -> CancellationToken {
        self.cancel.clone()
    }

    /// Handles incoming opus frames from the server.
    pub async fn handle_opus_frame(&self, frame: StampedOpusFrame) {
        if let Err(e) = self.opus_frames_tx.send(frame).await {
            warn!("failed to buffer opus frame: {}", e);
        }
    }

    /// Handles incoming commands from the server.
    pub async fn handle_command(&self, cmd: SessionCommandEvent) {
        let closed = self.closed.read().await;
        if *closed {
            return;
        }

        if let Err(e) = self.commands_tx.send(cmd).await {
            warn!("commands channel full, dropping command: {}", e);
        }
    }

    /// Receives data from the given DownlinkRx until closed.
    /// This spawns background tasks that will process incoming data.
    pub async fn recv_from<R: DownlinkRx + 'static>(&self, rx: Arc<R>) {
        let cancel = self.cancel.clone();
        let opus_tx = self.opus_frames_tx.clone();
        let cmd_tx = self.commands_tx.clone();
        let closed = self.closed.clone();

        // Spawn opus frames receiver
        let rx_opus = rx.clone();
        let cancel_opus = cancel.clone();
        tokio::spawn(async move {
            loop {
                tokio::select! {
                    _ = cancel_opus.cancelled() => break,
                    result = rx_opus.recv_opus_frame() => {
                        match result {
                            Ok(Some(frame)) => {
                                if opus_tx.send(frame).await.is_err() {
                                    break;
                                }
                            }
                            Ok(None) => break,
                            Err(_) => break,
                        }
                    }
                }
            }
        });

        // Spawn commands receiver
        let rx_cmd = rx.clone();
        let cancel_cmd = cancel.clone();
        tokio::spawn(async move {
            loop {
                tokio::select! {
                    _ = cancel_cmd.cancelled() => break,
                    result = rx_cmd.recv_command() => {
                        match result {
                            Ok(Some(cmd)) => {
                                let is_closed = *closed.read().await;
                                if is_closed {
                                    break;
                                }
                                if cmd_tx.send(cmd).await.is_err() {
                                    break;
                                }
                            }
                            Ok(None) => break,
                            Err(_) => break,
                        }
                    }
                }
            }
        });
    }

    /// Closes the port.
    pub async fn close(&self) -> Result<(), PortError> {
        self.cancel.cancel();
        *self.closed.write().await = true;
        self.tx.close().await.map_err(|e| PortError::Other(e.to_string()))
    }
}

#[async_trait]
impl<T: UplinkTx + 'static> ClientPortTx for ClientPort<T> {
    async fn send_opus_frame(&self, frame: &StampedOpusFrame) -> Result<(), PortError> {
        self.tx
            .send_opus_frame(frame)
            .await
            .map_err(|e| PortError::SendFailed(e.to_string()))
    }

    async fn send_state(&self, state: &GearStateEvent) -> Result<(), PortError> {
        self.tx
            .send_state(state)
            .await
            .map_err(|e| PortError::SendFailed(e.to_string()))
    }

    async fn send_stats(&self, stats: &GearStatsEvent) -> Result<(), PortError> {
        self.tx
            .send_stats(stats)
            .await
            .map_err(|e| PortError::SendFailed(e.to_string()))
    }
}

#[async_trait]
impl<T: UplinkTx + 'static> ClientPortRx for ClientPort<T> {
    async fn recv_opus_frame(&self) -> Option<StampedOpusFrame> {
        self.opus_frames_rx.lock().await.recv().await
    }

    async fn recv_command(&self) -> Option<SessionCommandEvent> {
        self.commands_rx.lock().await.recv().await
    }
}

#[cfg(test)]
mod client_port_tests {
    use super::*;
    use crate::{conn_pipe::new_pipe, GearState, UplinkRx};

    #[tokio::test]
    async fn test_client_port_send_state() {
        let (server, client) = new_pipe();
        let port = ClientPort::new(client);

        let state = GearStateEvent::new(GearState::Recording);
        port.send_state(&state).await.unwrap();

        let received = server.recv_state().await.unwrap().unwrap();
        assert_eq!(received.state, GearState::Recording);
    }

    #[tokio::test]
    async fn test_client_port_send_stats() {
        let (server, client) = new_pipe();
        let port = ClientPort::new(client);

        let mut stats = GearStatsEvent::new();
        stats.volume = Some(crate::Volume {
            percentage: 50.0,
            update_at: giztoy_jsontime::Milli::now(),
        });
        port.send_stats(&stats).await.unwrap();

        let received = server.recv_stats().await.unwrap().unwrap();
        assert_eq!(received.volume.unwrap().percentage, 50.0);
    }

    #[tokio::test]
    async fn test_client_port_send_opus() {
        let (server, client) = new_pipe();
        let port = ClientPort::new(client);

        let frame = StampedOpusFrame::now(vec![0x01, 0x02, 0x03]);
        port.send_opus_frame(&frame).await.unwrap();

        let received = server.recv_opus_frame().await.unwrap().unwrap();
        assert_eq!(received.frame, frame.frame);
    }

    #[tokio::test]
    async fn test_client_port_handle_command() {
        let (_server, client) = new_pipe();
        let port = ClientPort::new(client);

        // Manually handle a command
        let cmd = crate::SetVolume::new(75);
        let event = crate::SessionCommandEvent::new(&cmd);
        port.handle_command(event.clone()).await;

        // Use the async trait method
        let received = port.recv_command().await.unwrap();
        assert_eq!(received.cmd_type, "set_volume");

        port.close().await.unwrap();
    }

    #[tokio::test]
    async fn test_client_port_close() {
        let (_server, client) = new_pipe();
        let port = ClientPort::new(client);
        port.close().await.unwrap();
    }
}
